packages/firestore/src/util/misc.ts

```diff
@@ -74,6 +75,62 @@ export interface Equatable<T> {
   isEqual(other: T): boolean;
 }
 
+/** Compare strings in UTF-8 encoded byte order */
+export function compareUtf8Strings(left: string, right: string): number {
+  let i = 0;
+  while (i < left.length && i < right.length) {
+    const leftCodePoint = left.codePointAt(i)!;
+    const rightCodePoint = right.codePointAt(i)!;
+
+    if (leftCodePoint !== rightCodePoint) {
+      if (leftCodePoint < 128 && rightCodePoint < 128) {
+        // ASCII comparison
+        return primitiveComparator(leftCodePoint, rightCodePoint);
+      } else {
+        // Lazy instantiate TextEncoder
+        const encoder = newTextEncoder();
+
+        // UTF-8 encode the character at index i for byte comparison.
+        const leftBytes = encoder.encode(getUtf8SafeSubstring(left, i));
+        const rightBytes = encoder.encode(getUtf8SafeSubstring(right, i));
+        for (
+          let j = 0;
+          j < Math.min(leftBytes.length, rightBytes.length);
+          j++
+        ) {
+          const comp = primitiveComparator(leftBytes[j], rightBytes[j]);
+          if (comp !== 0) {
+            return comp;
+          }
+        }
```

Comment:
IIUC we're having this loop because we can't use `Buffer.from(...)` in the browser (outside of Node). If `leftBytes.length` is not equal to `rightBytes.length` and the first `Math.min(...)` bytes do match, then this loop won't return a comparison. So the code falls through to the `primitiveComparator` (line 112) way more often than it should.

- can we use the `compareBlobs` function?
- if not, can you take this loop out of here and make a helper function that compares 2 Uint8Arrays byte by byte and also considers the length of them?